lib/elixir/lib/module/types/descr.ex

```diff
@@ -1850,7 +1859,8 @@ defmodule Module.Types.Descr do
     end
   end
 
-  defp tuple_union(left, right), do: left ++ right
+  # Removes duplicates in union, which should trickle to other operations.
+  defp tuple_union(left, right), do: left ++ (right -- left)
```

Comment:
This is performing structural equality. Should we add a TODO saying this is a cheap optimization?